### DISCUSSION TITLE
CNP to report Disk usage

### DISCUSSION THREAD

**Discussion by dhilipkumars** [cloudnative-pg/cloudnative-pg#495] [Ideas]
## Problem:
Its not easy or straight forward to gather disk-usage on existing pg clusters. 

## Feature Request
Can cloudnative-pg objects report disk usage for every `x%` increment in the usage?  For eg: report every 10% increment of diskusage. 

## Proposed solution:
A possible option could be have the instance manager (side-car container) measure the disk-usage at a regular interval and report the usage under `status.Condition` ?

cc: @gbartolini @mnencia

**Comment by dhilipkumars** [comment]
this can be a building block to support feature requests like [these](https://github.com/cloudnative-pg/cloudnative-pg/issues/373)

**Comment by phisco** [comment]
kubelet is already reporting disk usage, see kubelet_volume_stats_used_bytes in https://github.com/cofyc/kubelet-exporter/blob/master/docs/metrics.md. This should be handled through a well configured monitoring stack, not the operator itself. Depending on the disk, periodically running a `du` could kill disk performances, I saw that happening on Jenkins with the disk usage plugin enabled.

**Reply by dhilipkumars** [reply]
Agreed `du` can be one rudimentary approach to solve this problem (especially only if we enable this as an optional beta feature && recommend this only for high performance disk), but i was actually looking for some ways that we can 1) query PG and ask about its disk-usage (if PG caches this value somewhere?), 2) @gbartolini also mentioned some interesting ideas on how we could potentially estimate database size by a query to catalog and WAL size based on the number of wal files on the disk.  3) CNP's instance manager directly consuming the kublet's `kubelet_volume_stats_used_bytes` stat is a great approach too IMO.  

CNP is already responsible for all the aspects of PG cluster, it manages end-to-end lifecycle of the the pg-clusters,  by reporting the disk-usage it improves the user experience massively as the user doesn't have to go and build complex system or integrate 3rd party controllers get get this info.

**Reply by phisco** [reply]
I don't think you can have a production-grade cluster without monitoring and Prometheus or Prometheus-compatible systems are the standard approaches to it. I would agree if the information pg could give was different from the one the kubelet reports given some additional knowledge it might have, but given that it would be exactly the same number, I'd say it's not in scope for the instance manager, nor the operator. I think that the Cluster resource should not be filled with information that can already be found somewhere else just because it's easier for POCs and demos, we should focus on production use cases.

**Reply by sxd** [reply]
I'm totally agree with @phisco, this is not work for the instance manager but for the external tools monitoring the Kubernetes cluster.

1. Definitely not a good idea, querying the disk for space usage every X seconds, it's not a good idea, and yes it will be every X seconds because it will run at the same time we ask for monitoring status
2. Don't know that idea, but sounds more like an heuristic rather than something to know for sure, but will like to read the full story there
3. If instance manager is going to read the `kubelet_volume_stats_used_bytes` isn't the same that you reading that from other place that will red the same stats from the operator? probably I'm misunderstanding something here

In my opinion, the operator should not be worry about the storage usage, that's part of the Kubernetes monitoring system, which actually knows about this, there's a ticket that will allow to put the WAL files in another volume, this will allow to exactly calculate the storage used by the database and the WAL files only by looking into the storage used by the database volume.

**Comment by sxd** [comment]
I'm turning this into a discussion since it's deserve a discussion before having any issue to work on

**Comment by Lillecarl** [comment]
```python
#! /usr/bin/env python3

from kubernetes import client, config, dynamic

# Configs can be set in Configuration class directly or using helper utility
config.load_kube_config()

v1 = client.CoreV1Api()

for i in v1.list_node().items:
    data_json: str = str(v1.connect_get_node_proxy_with_path(name=i.metadata.name, path='stats/summary'))

    # wtf https://sourcegraph.com/github.com/devopstales/KubeDash@b5879fbd9c676b7b3397ff1f9dafbfa3a899b8a1/-/blob/src/kubedash/functions/k8s.py?L563
    data_dict = eval(data_json)
    for pod in (data_dict["pods"] if "pods" in data_json else []):
        for volume in (pod["volume"] if "volume" in pod else []):
            if "pvcRef" in volume:
                print({
                    "name": volume['pvcRef']['name'],
                    "capacityBytes": int(volume['capacityBytes'])/1024,
                    "usedBytes": int(volume['usedBytes'])/1024,
                    "availableBytes": int(volume['availableBytes'])/1024,
                    "percentageUsed": (volume['usedBytes'] / volume['capacityBytes']  * 100),
                })
```

By proxying requests from the apiserver to the kubelet we can gather information about volume disk usage in a clean way[1]. We know where our pods are running so we don't need to query every node, so scalability works.

[Autogrowing](https://github.com/cloudnative-pg/cloudnative-pg/discussions/494) very much aligns with "Level V - Auto Pilot"

1: I don't know why I'm running eval on a json string, stolen from [KubeDash](https://sourcegraph.com/github.com/devopstales/KubeDash@b5879fbd9c676b7b3397ff1f9dafbfa3a899b8a1/-/blob/src/kubedash/functions/k8s.py?L563)

**Reply by Lillecarl** [reply]
```log
infra [ main][$!?][ v3.11.9][  /][🐚fish][⏱ 2s]
[15:32:31]❌5 ❯ ./volumes.py | rg "db-"
{'name': 'kronosdb-5-wal', 'capacityBytes': 10218772.0, 'usedBytes': 589896.0, 'availableBytes': 9612492.0, 'percentageUsed': 5.772670140795783}
{'name': 'kronosdb-5', 'capacityBytes': 102626232.0, 'usedBytes': 3619520.0, 'availableBytes': 98990328.0, 'percentageUsed': 3.5268955407034723}
{'name': 'grafanadb-3', 'capacityBytes': 20466256.0, 'usedBytes': 633448.0, 'availableBytes': 19816424.0, 'percentageUsed': 3.09508490463522}
{'name': 'keycloakdb-2', 'capacityBytes': 10218772.0, 'usedBytes': 697912.0, 'availableBytes': 9504476.0, 'percentageUsed': 6.829705173968066}
{'name': 'galacticdb-8', 'capacityBytes': 102626232.0, 'usedBytes': 33259880.0, 'availableBytes': 69349968.0, 'percentageUsed': 32.408751010170576}
{'name': 'keycloakdb-1', 'capacityBytes': 10218772.0, 'usedBytes': 697520.0, 'availableBytes': 9504868.0, 'percentageUsed': 6.825869096599865}
{'name': 'grafanadb-2', 'capacityBytes': 20466256.0, 'usedBytes': 633600.0, 'availableBytes': 19816272.0, 'percentageUsed': 3.0958275905470938}
{'name': 'galacticdb-8668', 'capacityBytes': 102626232.0, 'usedBytes': 33225428.0, 'availableBytes': 69384420.0, 'percentageUsed': 32.37518064582163}
{'name': 'galacticdb-8667', 'capacityBytes': 102626232.0, 'usedBytes': 33225468.0, 'availableBytes': 69384380.0, 'percentageUsed': 32.37521962221121}
{'name': 'kronosdb-7', 'capacityBytes': 102626232.0, 'usedBytes': 3609200.0, 'availableBytes': 99000648.0, 'percentageUsed': 3.5168396321907247}
{'name': 'kronosdb-7-wal', 'capacityBytes': 10218772.0, 'usedBytes': 573504.0, 'availableBytes': 9628884.0, 'percentageUsed': 5.612259476970423}
{'name': 'keycloakdb-3', 'capacityBytes': 10218772.0, 'usedBytes': 697748.0, 'availableBytes': 9504640.0, 'percentageUsed': 6.828100284456879}
{'name': 'kronosdb-6-wal', 'capacityBytes': 10218772.0, 'usedBytes': 573504.0, 'availableBytes': 9628884.0, 'percentageUsed': 5.612259476970423}
{'name': 'kronosdb-6', 'capacityBytes': 102626232.0, 'usedBytes': 3609188.0, 'availableBytes': 99000660.0, 'percentageUsed': 3.51682793927385}
{'name': 'grafanadb-4', 'capacityBytes': 20466256.0, 'usedBytes': 633440.0, 'availableBytes': 19816432.0, 'percentageUsed': 3.095045815903016}
```
As you can see we have slightly overprovisioned our (dev) clusters 😄
